deployment/cfn/application.py

```diff
@@ -0,0 +1,363 @@
+            Description='Application server AutoScalingGroup maximum'
+        ), 'AppServerAutoScalingMax')
+
+        self.ssl_certificate_arn = self.add_parameter(Parameter(
+            'SSLCertificateARN', Type='String',
+            Description='ARN for a SSL certificate stored in IAM'
+        ), 'SSLCertificateARN')
+
+        self.public_subnets = self.add_parameter(Parameter(
+            'PublicSubnets', Type='CommaDelimitedList',
+            Description='A list of public subnets'
+        ), 'PublicSubnets')
+
+        self.private_subnets = self.add_parameter(Parameter(
+            'PrivateSubnets', Type='CommaDelimitedList',
+            Description='A list of private subnets'
+        ), 'PrivateSubnets')
+
+        self.vpc_id = self.add_parameter(Parameter(
+            'VpcId', Type='String',
+            Description='VPC ID'
+        ), 'VpcId')
+
+        self.notification_topic_arn = self.add_parameter(Parameter(
+            'GlobalNotificationsARN', Type='String',
+            Description='ARN for an SNS topic to broadcast notifications'
+        ), 'GlobalNotificationsARN')
+
+        app_server_lb_security_group, app_server_security_group = self.create_security_groups()  # NOQA
+        app_server_lb = self.create_load_balancer(app_server_lb_security_group)
+
+        self.create_auto_scaling_resources(app_server_security_group,
+                                           app_server_lb)
+
+        self.create_cloud_watch_resources(app_server_lb)
+
+        self.add_output(Output('AppServerLoadBalancerEndpoint',
+                               Value=GetAtt(app_server_lb, 'DNSName')))
+        self.add_output(Output('AppServerLoadBalancerHostedZoneNameID',
+                               Value=GetAtt(app_server_lb,
+                                            'CanonicalHostedZoneNameID')))
+
+    def get_recent_application_server_ami(self):
+        try:
+            application_server_ami_id = self.get_input('AppServerAMI')
+        except MKUnresolvableInputError:
+            application_server_ami_id = get_recent_ami(self.aws_profile,
+                                                       'mmw-app-*',
+                                                       owner='self')
+
+        return application_server_ami_id
+
+    def create_security_groups(self):
+        app_server_lb_security_group_name = 'sgAppServerLoadBalancer'
+
+        app_server_lb_security_group = self.add_resource(ec2.SecurityGroup(
+            app_server_lb_security_group_name,
+            GroupDescription='Enables access to application servers via a '
+                             'load balancer',
+            VpcId=Ref(self.vpc_id),
+            SecurityGroupIngress=[
+                ec2.SecurityGroupRule(
+                    IpProtocol='tcp', CidrIp=ALLOW_ALL_CIDR, FromPort=p,
+                    ToPort=p
+                )
+                for p in [HTTP, HTTPS]
+            ],
+            SecurityGroupEgress=[
```

Comment:
Under the covers somewhere they might, but none of that is exposed to us. The reason why HTTP is enabled for egress here is because ELBs send health checks to the application servers via HTTP. For ingress, HTTP and HTTPS are enabled because we accept HTTP traffic, but then redirect users to the same URL over HTTPS.